### ISSUE TITLE
Replacement for the old DebugAPI.json files

### ISSUE THREAD

**Issue by jaregier** [apollographql/apollo-android#1748]
We recently tried upgrading our Apollo to use the latest compiler.  What we found was that a file that was generated as part of the old compiler, `DebugAPI.json` (or `ReleaseAPI.json`, whatever build variant you are running) is no longer generated as part of the codegen step.  I understand that the new compiler does things differently, so let me explain the use case:

We use persisted queries, but we don't use [automatic persisted queries](https://www.apollographql.com/docs/apollo-server/performance/apq/).  We manually sync the persisted queries to our server as part of a CI step.  Previously, we would use the `DebugAPI.json` and send it to our server, but really all we needed was a mapping of operationId -> sourceWithFragments.

Now that this file is no longer generated, we'll need to find another way to get this mapping.  We aren't seeing any mapping files generated as part of the build process; we've also tried using the [extract](https://github.com/apollographql/apollo-tooling#apollo-clientextract-output) command from the CLI interface, but differences in spacing and sorting seem to be causing it to generate signatures differently from what Apollo is generating for our Operation IDs.

TLDR: Is there a way for us to generate a json file that maps the operation IDs to the sourceWithFragments with the new compiler?

**Comment by martinbonnin** [comment]
You can look into `transformedQueries`: https://github.com/apollographql/apollo-android/issues/1597

If I understand correctly, this was added specifically for this use case.

**Comment by jaregier** [comment]
We tried with the transformed queries, and while they do output the actual queries that Apollo uses under the hood, there's no Operation ID output to go with them (at least from what I can see).  We were hoping we could use those queries and run them through Apollo CLI to get the Operation ID, but the CLI seems to have slightly different logic before it generates the signature (different whitespace between the two and it also seems to be sorting the fragments), so when the sha256 is run on them, they don't match.

@DSteve595 Are you approaching persisted queries in a similar way?  If so, did you run into this same hash problem?

**Comment by DSteve595** [comment]
In our case, the backend provides operation IDs for given queries. We ignore the operation ID created by the compiler, for the reasons you mentioned. Is your backend able to provide them?

**Comment by martinbonnin** [comment]
@DSteve595 , if the server returns a serverID, somehow the client needs to remember this and map from operationId to serverId, right ? Or a way to go from MyQuery to serverId ? I can see how someone could just use apollo-runtime and just strip the operation while using OperationID in all cases.

I was going to suggest we make the transformed query match exactly what is send over the wire in all cases (including minification and sorting the fragments) ? That will not be very human-readable but I guess it's ok ?

**Comment by jaregier** [comment]
We could even just have an additional json file that gets output alongside the existing transformed queries.  The `manifest.json` that is generated by [CLI extract](https://github.com/apollographql/apollo-tooling#apollo-clientextract-output) is a good and simple format, so if we had a similar file but just use the apollo-android operation IDs as the signature, that would probably be ideal for us.  This would be similar to how we used the DebugAPI.json file in the past, although we might be the only ones that relied on that file given the lack of issues.

Getting the backend to generate the hash is something we've done in the past, but it's cumbersome as it means we need to commit the persisted query hashes, whereas using the apollo Operation ID means we can just generate them at compile time, and then upload the persisted queries after the fact.  It's easier to hook into CI, and it also means any issues with forgetting to sync persisted queries can be fixed without doing a full release.

**Comment by DSteve595** [comment]
We do that server ID fetching at build time, not run time. We map (with codegen) each Apollo-generated operation class to a server ID.

**Comment by DSteve595** [comment]
>make the transformed query match exactly what is sent over the wire in all cases

This makes sense IMO

**Comment by martinbonnin** [comment]
@DSteve595 Pull Request there to make `transformedQuery.readText() == operation.queryDocument()`. Let me know if it works for you.

**Comment by martinbonnin** [comment]
@jaregier can you check the new transformed queries ? You should be able to run a sha256 on them to get the id.

**Comment by jaregier** [comment]
@martinbonnin Yeah, this works for us, it's able to generate the same sha256 now.  Thanks for the help, appreciate it!
